### ISSUE TITLE
The most important operations runs in main thread.

### ISSUE THREAD

**Issue by Leta0n** [Haneke/Haneke#69]
I saw some lags in my app when i started to use Heneke. And i found this problem. Its happens because your fetchers perform their success blocks in main thread.

Some proves:
![screen shot 2015-03-01 at 19 47 36](https://cloud.githubusercontent.com/assets/3305500/6432056/ef9bd18c-c04c-11e4-9418-730716176861.png)
![screen shot 2015-03-01 at 19 47 10](https://cloud.githubusercontent.com/assets/3305500/6432057/ef9dd234-c04c-11e4-9924-4787d4938df9.png)
![screen shot 2015-03-01 at 19 51 21](https://cloud.githubusercontent.com/assets/3305500/6432059/efa07926-c04c-11e4-9c01-362e12f7831e.png)
![screen shot 2015-03-01 at 19 47 28](https://cloud.githubusercontent.com/assets/3305500/6432058/ef9fbf54-c04c-11e4-973b-558b39bc9aea.png)

**Comment by raulrea** [comment]
Having this issue as well. 
Anyone tried wrapping any of those methods in a separate queue?

**Comment by bdorfman** [comment]
We have done a lot of work on the Foursquare fork to move everything off to other threads. Haven't had a chance to clean everything up and submit back to this repo yet though. 

You could try using our version if you like, but no guarantees that things won't break or method names won't change.

**Comment by raulrea** [comment]
Great!

I’ll check it out.

https://github.com/foursquare/Haneke <https://github.com/foursquare/Haneke> right?

**Comment by bdorfman** [comment]
Yep. We also have some other improvements, like the ability to load images synchronously when there is a cache hit, consolidating duplicate in-flight network requests, and the ability to set what network fetcher subclass is used in the built in UIKit category methods so you can provide a custom URLSession.

**Comment by raulrea** [comment]
Awesome!
Giving it a try right now.

**Comment by skydivedan** [comment]
Would you consider making a cocoapod for your fork? There are some pretty nice changes in here.

**Comment by bdorfman** [comment]
I think the idea is to submit our changes back to the main repo once we have some free time to clean things up a bit. Hopefully in the next month or so. I'd much rather do that than start "officially" maintaining a separate fork.

**Comment by houmie** [comment]
@bdorfman It would be good to see these changes in main repo. Although it seems Haneke is no longer active.
Do you guys at foursquare/Haneke have a way to prefetch images in an array? Like SDWebImage does?

```
[[SDWebImagePrefetcher sharedImagePrefetcher] prefetchURLs:<NArray with image URLs>];
```

Then I would be using your fork right away.
